server/src/com/cloud/storage/StorageManagerImpl.java

```diff
@@ -1104,14 +1107,61 @@ public void cleanupStorage(boolean recurring) {
+                                        if ((deleteSnapshotsDirAnswer != null) && deleteSnapshotsDirAnswer.getResult()) {
+                                            s_logger.debug("Deleted snapshot: " + snapshotVO.getUuid() + " from secondary storage: " + snapshotDataStore.getName());
+                                            _snapshotStoreDao.expunge(ref.getId());
+                                        } else {
+                                            s_logger.warn("Failed to delete snapshot: " + snapshotVO.getUuid() + " from secondary storage: " + snapshotDataStore.getName());
+                                            removeSnapshot = false;
+                                        }
+                                    } else if (ref.getRole() == DataStoreRole.Primary) { // Delete worker VM snapshot
+                                        VolumeVO volume = _volumeDao.findByIdIncludingRemoved(snapshotVO.getVolumeId());
+                                        if (volume.getInstanceId() == null)
+                                            continue;
+                                        VMInstanceVO vm = _vmInstanceDao.findById(volume.getInstanceId());
+                                        Long hostId = vm.getHostId() != null ? vm.getHostId() : vm.getLastHostId();
+                                        if (hostId != null) {
+                                            VMSnapshotTO vmSnapshotTO = new VMSnapshotTO();
+                                            vmSnapshotTO.setSnapshotName(ref.getInstallPath());
+                                            vmSnapshotTO.setDescription(snapshotVO.getName());
+                                            DeleteVMSnapshotCommand deleteSnapshotCommand = new DeleteVMSnapshotCommand(vm.getInstanceName(), vmSnapshotTO, null, null);
+                                            Answer deleteSnapshotAnswer = _agentMgr.send(hostId, deleteSnapshotCommand);
+                                            if ((deleteSnapshotAnswer != null) && deleteSnapshotAnswer.getResult()) {
+                                                s_logger.debug("Deleted worker VM snapshot: " + snapshotVO.getName());
+                                                _snapshotStoreDao.expunge(ref.getId());
+                                            } else {
+                                                s_logger.warn("Failed to delete worker VM snapshot: " + snapshotVO.getName());
+                                                removeSnapshot = false;
+                                            }
+                                        }
+                                    }
+                                } else {
+                                    _snapshotStoreDao.expunge(ref.getId());
+                                }
+                            }
+                            if (removeSnapshot) {
+                                _snapshotDao.expunge(snapshotVO.getId());
```

Comment:
this method is already to long.
please factor out the new part into a seperate method.